### DISCUSSION TITLE
graphql.language.Document - schema definition usage

### DISCUSSION THREAD

**Discussion by adarsh0048** [graphql-java/graphql-java#2813] [Q&A]
Hi, 
Thanks for reading this !

I have used the class IntrospectionResultToSchema to convert the response from introspection query to a Document - 

`IntrospectionResultToSchema introspectionResultToSchema = new IntrospectionResultToSchema();
            Object ob = introspectionResultToSchema.createSchemaDefinition(schemaMap);`

Now, I wanted to understand, which other operations can I do on this result(ob) Document instance using this repo. I converted it so that now, the response has some structure rather than the entire json response from the introspection query. 
My aim, is to convert it into a json or form a query, with some fields selected from all of the fields in the response. If you can direct me to some documentation or a class, that would be helpful. 

Best,
Adarsh

**Comment by bbakerman** [comment]
`IntrospectionResultToSchema`  can be used to re-create a schema that came from a live system.

When you throw a full schema [introspection query](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/introspection/IntrospectionQuery.java#L8) at a running graphql server you get back JSON that represents the schema.

The `IntrospectionResultToSchema` class helps turn that back into SDL

eg

```
type Query {
   feild1 : Type1
}

type Type1 {
   foo : String
}
```

(except its parsed already and in Document runtime AST form.

You could use `graphql.schema.idl.SchemaPrinter` to print it back out as SDL text.

You have said

> My aim, is to convert it into a json or form a query, with some fields selected from all of the fields in the response.

This is the bit I am unclear on - can you explain more clearly what you want to achieve (think less about the how and more about the outcome is you want to achieve is)

For example do you want to build a new schema from the introspection and then change what graphql types and fields are in that new schema?

**Reply by adarsh0048** [reply]
Thanks [bbakerman](https://github.com/bbakerman) for taking time to answer. 

What I am trying to achieve is something like what Github explorer is doing. https://docs.github.com/en/graphql/overview/explorer

So, here if someone wants to select a few fields (click on Explorer, next to Prettify and History) , they can see which fields are available, then choose. Once it's done, automatically a query is formed, which can be sent to the server to the get the output.

 
![Screen Shot 2022-04-29 at 10 11 13 AM](https://user-images.githubusercontent.com/90865574/165991533-bdb7d5c5-fde6-4e65-b6a5-e5b30f7d8f30.png)


So, first to have that menu or catalog, where the fields are being selected I need to have the details about all the introspection response json, so I was wondering if there was a way to get it in some structured way. So, I found that `IntrospectionResultToSchema introspectionResultToSchema = new IntrospectionResultToSchema(); Object ob = introspectionResultToSchema.createSchemaDefinition(schemaMap);` was converting the whole introspection response json into a Document object.

Now, I want to utilize this Document object, to convert to some form (maybe that is SDL text) like a Map, nested List or some tree like structure, so that I can display all the fields, their sub fields, types etc available for selection. Then once selected, I want to be able to convert those to a query to get the actual output.

Let me know if you are able to picture what I am trying to do or else I can explain more.
